src/diffusers/models/transformers/transformer_hidream_image.py

```diff
@@ -663,12 +663,24 @@ def __init__(
 
     def unpatchify(self, x: torch.Tensor, img_sizes: List[Tuple[int, int]], is_training: bool) -> List[torch.Tensor]:
         if is_training:
-            B, S, F = x.shape
-            C = F // (self.config.patch_size * self.config.patch_size)
-            x = (
-                x.reshape(B, S, self.config.patch_size, self.config.patch_size, C)
-                .permute(0, 4, 1, 2, 3)
-                .reshape(B, C, S, self.config.patch_size * self.config.patch_size)
+            pH, pW = img_sizes[0]  # Get target patch height/width
+            expected_S = pH * pW
+            # Ensure sequence length S matches expected H*W before rearranging
+            # This might require padding/truncating x if training uses fixed max_seq
+            current_S = x.shape[1]
+            if current_S > expected_S:
+                x = x[:, :expected_S, :]  # Use only the relevant part of the sequence
+            elif current_S < expected_S:
+                # This case is less likely if padding happens earlier, but handle defensively
+                raise ValueError(
+                    f"Sequence length {current_S} is less than expected {expected_S} ({pH}x{pW}) during unpatchify.")
+
+            # Corrected line using einops and H, W:
```

Comment:
Just for the record, we cannot be using `einops` here.

Reply:
i think it is inherited from the original codebase, easy to overlook after installing einops for testing and then it works everywhere after :see_no_evil: